ostd/src/arch/riscv/irq/chip/mod.rs

```diff
@@ -0,0 +1,193 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! Interrupts.
+
+mod plic;
+
+use alloc::boxed::Box;
+use core::{
+    fmt,
+    ops::{Deref, DerefMut},
+};
+
+use spin::Once;
+
+use crate::{
+    arch::{
+        boot::DEVICE_TREE,
+        irq::{chip::plic::Plic, HwIrqLine, InterruptSource},
+    },
+    cpu::CpuId,
+    io::IoMemAllocatorBuilder,
+    irq::IrqLine,
+    sync::SpinLock,
+    Result,
+};
+
+/// The [`IrqChip`] singleton.
+pub static IRQ_CHIP: Once<IrqChip> = Once::new();
+
+/// Initializes the Platform-Level Interrupt Controller (PLIC).
+///
+/// # Safety
+///
+/// This function must be called once and at most once at a proper timing in the
+/// boot context.
+pub(in crate::arch) unsafe fn init(io_mem_builder: &IoMemAllocatorBuilder) {
+    let device_tree = DEVICE_TREE.get().unwrap();
+    let mut plics = Plic::from_fdt(device_tree, io_mem_builder);
+    plics.iter_mut().for_each(|plic| plic.init());
+    IRQ_CHIP.call_once(|| IrqChip {
+        plics: SpinLock::new(plics.into_boxed_slice()),
+    });
+    // SAFETY: Accessing the `sie` CSR to enable the external interrupt is
+    // safe here because this function is only called during PLIC
+    // initialization, and we ensure that only the external interrupt bit is
+    // set without affecting other interrupt sources.
+    unsafe { riscv::register::sie::set_sext() };
+}
+
+/// An IRQ chip.
+///
+/// This abstracts the hardware IRQ chips (or IRQ controllers), allowing the bus
+/// or device drivers to enable [`IrqLine`]s (via, e.g., [`map_interrupt_source_to`])
+/// regardless of the specifics of the IRQ chip.
+///
+/// In the RISC-V architecture, the underlying hardware is typically Platform-Level
+/// Interrupt Controller (PLIC).
+///
+/// [`map_interrupt_source_to`]: Self::map_interrupt_source_to
+pub struct IrqChip {
+    plics: SpinLock<Box<[Plic]>>,
```

Comment:
Use `SpinLock<..., LocalIrqDisabled>`. Otherwise there is dead lock if IRQ comes in `map_interrupt_source_to`.